File: app/src/main/java/ru/example/newsreader/MainActivity/MainActivityPresenter.java
/*
package ru.example.newsreader.MainActivity;

import android.content.Context;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import ru.example.newsreader.UtilsKt;
import ru.example.newsreader.models.Article;
import ru.example.newsreader.models.RSSFeed;
import ru.example.newsreader.retrofit.RetrofitClientKt;
import ru.example.newsreader.room.AppDatabase;
import ru.example.newsreader.room.Entity.ArticleEntity;

public class MainActivityPresenter {
    private MainActivityView mainActivityView;
    private Context context;
    private AppDatabase appDatabaseKt;

    MainActivityPresenter(MainActivityView mainActivityView, AppDatabase appDatabaseKt, Context context){
        this.mainActivityView=mainActivityView;
        this.appDatabaseKt = appDatabaseKt;
        this.context = context;
    }

    public void downloadArticles(){
        if (UtilsKt.Companion.hasConnection(context)) {
            new RetrofitClientKt().getArticles().enqueue(new Callback<RSSFeed>() {
                @Override
                public void onResponse(Call<RSSFeed> call, Response<RSSFeed> response) {
                    mainActivityView.showArticles(response.body().getArticleList());
                    new Thread(() -> {
                        if(getArticlesFromDB().size()>0){}
                        else saveArticles(response.body().getArticleList());
                    }).start();
                }
                @Override
                public void onFailure(Call<RSSFeed> call, Throwable t) {
                }
            });
        }
        else{
            new Thread(() -> {
                try {
                    mainActivityView.showArticles(getArticlesFromDB());
                } catch (Exception e) {
                }
            }).start();
        }
    }

    private void saveArticles(List<Article> articles){
        Thread t = new Thread() {
            public void run() {
                for (Article article: articles) {
                    appDatabaseKt.getArticleDao().insert(article.convertToArticleEntity());
                }
            }
        };
        t.start();

    }

    private List<Article> getArticlesFromDB(){
        List<ArticleEntity> articleEntities = appDatabaseKt.getArticleDao().getAllArticles();
        List<Article> articleList = new ArrayList<>();
        for (ArticleEntity articleEntity: articleEntities) {
            articleList.add(articleEntity.convertToArticle());
        }
        return articleList;
    }

    public void deleteArticlesFromDB(){
        new Thread(() -> {
            appDatabaseKt.getArticleDao().deleteAll();
            try {
                appDatabaseKt.getArticleDao().deleteAll();
            }
            catch (Exception e) {

            }
        }).start();
    }

    public void getArticlesInSource(){
        new Thread(() -> {
            try {
                //List<ArticleEntity> articleEntities = appDatabase.getArticleDao().findArticlesForSource("habr.com");
            }
            catch (Exception e) {
            }
        }).start();
    }
}
*/

File: app/src/main/java/ru/example/newsreader/screens/main_activity/MainActivityPresenter.kt
package ru.example.newsreader.screens.main_activity

import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.LifecycleObserver
import android.arch.lifecycle.OnLifecycleEvent
import android.content.Context
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.plusAssign
import ru.example.newsreader.models.ArticleKt
import ru.example.newsreader.room.entity.ArticleEntity
import ru.example.newsreader.utils.Utils

interface MainActivityPresenter{
    fun getArticles()
    fun getArticlesFromDB(): Observable<List<ArticleEntity>>
    fun saveArticles(articles: MutableList<ArticleKt>)
    fun deleteArticlesFromDB()

    fun detachView()
    fun attachView(view: MainActivityView)
}

class MainActivityPresenterImpl (private val interactor: MainActivityInteractor,
                                 val context: Context) : LifecycleObserver, MainActivityPresenter {

    private var view: MainActivityView? = null
    private val disposables = CompositeDisposable()

    override fun attachView(view: MainActivityView) {
        this.view = view
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    override fun getArticles() {
        view?.showProgressBar()
        if (Utils.hasConnection(context)){
            disposables += interactor.downloadArticles()
                    .subscribe { response ->
                        view?.hideProgressBar()
                        view?.showArticles(response.articleList!!)
                        saveArticles(response.articleList)
            }
        }
        else {
            disposables += getArticlesFromDB().subscribe {
                view?.hideProgressBar()
                view?.showArticles(Utils.convertArticleEntityToArticle(it))
            }
        }
    }

    override fun getArticlesFromDB(): Observable<List<ArticleEntity>> {
        return interactor.getArticlesFromDB()
    }

    override fun saveArticles(articles: MutableList<ArticleKt>) {
        interactor.saveArticles(articles)
    }

    override fun deleteArticlesFromDB() {
        interactor.deleteArticlesFromDB()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    override fun detachView() {
        disposables.dispose()
        view = null
    }
}



